File: build_html.py
from glob import glob
import argparse
import chevron
import math
import os
import sys


def get_page_count(panel_count, panels_per_page):
    return math.ceil(panel_count / panels_per_page)


def get_page_index(i, panels_per_page):
    return math.floor(i / panels_per_page)


# [1-18] ->  [1,4,7,10,13,16,2,5,8,11,14,17,3,6,9,12,15,18]
def panelize(input, panels_per_page):
    output = []
    page_count = get_page_count(len(input), panels_per_page)

    for i in range(page_count * panels_per_page):
        page_index = get_page_index(i, panels_per_page)
        panel_index = i % panels_per_page
        input_index = panel_index * page_count + page_index

        appendee = input[input_index] if input_index <= len(input) - 1 else None

        output.append(appendee)

    return output


def get_panels(directory, panels_per_page):
    filenames = glob(directory + "/*.png")

    return [
        {
            "filename": os.path.relpath(filename, directory) if filename else None,
            "page": get_page_index(i, panels_per_page),
        }
        for i, filename in enumerate(panelize(sorted(filenames), panels_per_page))
    ]


def get_html(
    directory,
    rows,
    columns,
    orientation,
    page_width,
    page_height,
    page_padding,
    panel_padding,
):
    panels = get_panels(directory, rows * columns)

    pages = []
    for i in range(get_page_count(len(panels), rows * columns)):
        pages.append(
            {"panels": list(filter(lambda panel: panel.get("page") == i, panels))}
        )

    return chevron.render(
        template,
        # TODO: tidy
        # TODO: decouple handle position from is_portrait
        {
            "pages": pages,
            "rows": rows,
            "columns": columns,
            "is_portrait": orientation == "portrait",
            "page_width": page_width,
            "page_height": page_height,
            "page_padding": page_padding,
            "panel_padding": panel_padding,
        },
    )


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--directory",
        type=str,
        required=True,
        help="path to images' folder",
    )

    # TODO: DRY against make.sh
    parser.add_argument("--rows", type=int, default=3, help="Panel rows per page")
    parser.add_argument("--columns", type=int, default=3, help="Panel columns per page")

    parser.add_argument(
        "--orientation",
        type=str,
        default="landscape",
        help="Print orientation: 'landscape' (default) or 'portrait'",
    )

    parser.add_argument("--page_padding", type=str, default=".25in")
    parser.add_argument("--panel_padding", type=str, default=".0625in")

    # TODO: parameterize handle/image size, panel order, image filter

    arguments = parser.parse_args()

    dir_path = os.path.dirname(os.path.realpath(__file__))

    if not os.path.isdir(arguments.directory):
        sys.exit("ERROR: " + arguments.directory + " directory does not exist")

    if arguments.orientation not in ["portrait", "landscape"]:
        sys.exit("ERROR: invalid --orientation " + arguments.orientation)

    with open(dir_path + "/template.mustache", "r") as template:
        output = open(arguments.directory + "/index.html", "w")
        output.write(
            get_html(
                directory=arguments.directory,
                rows=arguments.rows,
                columns=arguments.columns,
                orientation=arguments.orientation,
                page_width="8.5in" if arguments.orientation == "portrait" else "11in",
                page_height="11in" if arguments.orientation == "portrait" else "8.5in",
                page_padding=arguments.page_padding,
                panel_padding=arguments.panel_padding,
            )
        )
        output.close()

File: build_html.ts
import { fileExtension } from "https://deno.land/x/file_extension@v2.1.0/mod.ts";
import { parse } from "https://deno.land/std@0.202.0/flags/mod.ts";
import { range } from "https://deno.land/x/lodash@4.17.15-es/lodash.js";
import { renderFile } from "https://deno.land/x/mustache@v0.3.0/mod.ts";

interface Panel {
  id: number;
  filename: string | undefined;
  page: number;
}

interface Page {
  panels: Panel[];
}

interface HtmlSettings {
  title: string;

  rows: number;
  columns: number;

  pageWidth: string;
  pageHeight: string;
  pagePadding: string;
  pageSide: "front" | "back";

  handlePadding: string;

  imageWidth: string;
  imageHeight: string;
  imageMargin: string;
  imagePosition: string;

  crop: true;
  imageFilter: string;

  flyleavesCount: number;
}

const DEFAULT_SETTINGS: HtmlSettings = {
  title: "output",

  rows: 5,
  columns: 2,

  pageWidth: "8.5in",
  pageHeight: "11in",
  pagePadding: ".5in .75in",
  pageSide: "front",

  handlePadding: ".125in",

  imageWidth: "2in",
  imageHeight: "1.875in",
  imageMargin: ".0625in",
  imagePosition: "center center",

  crop: true,
  imageFilter: "none",

  flyleavesCount: 2,
};

const getPageCount = (panelCount: number, panelsPerPage: number) =>
  Math.ceil(panelCount / panelsPerPage);

const getPageIndex = (i: number, panelsPerPage: number) =>
  Math.floor(i / panelsPerPage);

const getPanels = async (
  directory: string,
  panelsPerPage: number,
  flyleavesCount: HtmlSettings["flyleavesCount"],
): Promise<Panel[]> => {
  let filenames: (string | undefined)[] = [];
  for await (const dirEntry of Deno.readDir(directory)) {
    if (fileExtension(dirEntry.name) == "png") {
      filenames.push(dirEntry.name);
    }
  }

  // NOTE: this assumes numbers are zero-padded!
  filenames = filenames.sort();

  filenames = [
    ...Array(flyleavesCount).fill(undefined),
    ...filenames,
    ...Array(flyleavesCount).fill(undefined),
  ];

  return filenames.map((
    filename: string | undefined,
    i: number,
  ) => ({
    id: i,
    filename,
    page: getPageIndex(i, panelsPerPage),
  }));
};

const getPages = (
  panels: Panel[],
  rows: HtmlSettings["rows"],
  columns: HtmlSettings["columns"],
  pageSide: HtmlSettings["pageSide"],
) => {
  const pages: Page[] = [];

  range(0, getPageCount(panels.length, rows * columns))
    .forEach(
      (i: number) => {
        pages.push({
          panels: panels.filter((panel: Panel) => panel.page == i),
        });
      },
    );

  if (pageSide == "back") {
    pages.reverse();
  }

  return pages;
};

// deno-lint-ignore no-explicit-any
const getSettings = (flags: any): HtmlSettings => ({
  ...DEFAULT_SETTINGS,
  ...flags,
});

const getHtml = async (
  directory: string,
  settings: HtmlSettings,
) => {
  return await renderFile("template.mustache", {
    pages: getPages(
      await getPanels(
        directory,
        settings.rows * settings.columns,
        settings.flyleavesCount,
      ),
      settings.rows,
      settings.columns,
      settings.pageSide,
    ),
    pageDirection: settings.pageSide == "front" ? "ltr" : "rtl",
    ...settings,
  });
};

(async () => {
  const flags = parse(Deno.args);
  const settings = getSettings(flags);

  await Deno.writeTextFile(
    flags.directory + "/index.html",
    await getHtml(flags.directory, settings),
  );
  await Deno.writeTextFile(
    flags.directory + "/settings.json",
    JSON.stringify(settings, null, 2),
  );
})();
